Add an example of a dynamic resource provider that adds a custom domain to an Azure CDN endpoint.

praneetloke opened the pull request.

mikhailshilkov commented:
This seems to be related to #306, although with a different focus. Is that something we could/should combine?

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+});
+
+// Create an Azure resource (Storage Account) so we can point the CDN endpoint to it.
+const storageAccount = new azure.storage.Account("storage-account", {
```
Storage account names can't have dashes in them

praneetloke replied:
Ah yes. This is the result of me trying to change the names I used in my personal project to something generic. I'll change this.

mikhailshilkov commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,231 @@
+      throw new Error("ARM credentials could not be detected.");
+    }
+
+    const credentials = await msRestAzure.loginWithServicePrincipalSecret(clientID, clientSecret, tenantID);
```
This fails when I run it from CLI:

> Checking env vars for ARM credentials.
> Error: ARM credentials could not be detected.: Error: ARM credentials could not be detected.

Shouldn't we do something like `const cred = await msRestAzure.AzureCliCredentials.create();` instead?

praneetloke replied:
The dynamic provider tries to use the Pulumi CLI's Azure config, and if not looks for environment vars. I haven't tried using `msRestAzure.AzureCliCredentials.create();` myself, to be honest. I did it this way since I typically use a Service Connection in Azure DevOps and don't use the Az CLI in any of my tasks.

mikhailshilkov replied:
The README suggests `az login`, so it should work. Also, that's probably what users will use while trying examples.

praneetloke replied:
Good catch! I have updated this to try and get the credentials from the CLI if the env vars are not set.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+    origins: [
+        {
+            name: "cdn-origin",
+            hostName: storageAccount.primaryBlobEndpoint,
```
This needs to be a hostname like `bla.blob.core.windows.net` not the full URL like `https://bla.blob.core.windows.net`. Consider using something like `storageAccount.primaryBlobEndpoint.apply(ep => url.parse(ep).hostname!)`. Applies to `originHostHeader` too.

praneetloke replied:
`primaryBlobEndpoint` actually does work for both of those properties. But if you think mapping the hostname would be better, I can change it.

mikhailshilkov replied:
Hmm, I'm getting errors running that... weird

praneetloke replied:
Ah sorry. I compared this with the wrong project of mine. In my personal project, I am using the function app's host name as the origin. When I translated this to use the storage account, I used the wrong property. I should have used `primaryBlobHost`.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+            // pointing to my-cdn-endpoint.azureedge.net.
+            // You would do that in your domain registrar's portal.
+            customDomainHostName: "mycompany.com",
+            customDomainName: "custom-domain",
```
Why do we need this property? Could we just use `name` internally instead?

praneetloke replied:
Yep. Absolutely. I can change the provider to use the name of the custom resource as the "name" of the custom domain as well. Also, even though it is sort of a display name the Azure portal doesn't seem to show it anywhere.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+            // The certificate is fully managed by Azure from provisioning
+            // to automatic renewal at no additional cost to you.
+            httpsEnabled: true,
+            resourceGroupName: resourceGroupName
```
Nit: missing commas here and in other places

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+// Create an Azure resource (Storage Account) so we can point the CDN endpoint to it.
+const storageAccount = new azure.storage.Account("storage-account", {
+    resourceGroupName: resourceGroup.name,
+    location: resourceGroup.location,
```
You shouldn't pass the location explicitly: it will be auto-filled from resource group properties.

praneetloke replied:
Ah gotcha. Removed.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+
+export const cdnEndpointUrl = pulumi.interpolate `https://${cdnEndpoint.hostName}`;
+
+pulumi.all([resourceGroup.name, cdnProfile.name, cdnEndpoint.name])
```
I think it would look more natural if we accepted `Input<T>` in `CustomDomainOptions` and then on the client side it would be just a constructor call without `pulumi.all`.

praneetloke replied:
Yes, I definitely agree. In fact, that is what I started with. But if you look closely at the formal parameters of the functions in my dynamic resource provider, I used the type `CustomDomainOptions` for `inputs` instead of `any`. That means I will need to deal with an `Input<T>` inside those functions, which I couldn't see a way to unwrap them there without returning a `pulumi.Output<T>`, which I couldn't find a cleaner way to transform. Things got really confusing there to be honest.

I see that in your static website example, you are retaining the type of the `inputs` formal param as `any` and are accepting an `Input<T>` for your provider input. I am not sure how when your `create()` is called, that you are getting a regular string instead of an `Input<T>`. In any case, I would prefer to have the `inputs` param typed concretely instead of having to treat it as an anonymous object.

mikhailshilkov replied:
Great point! I stitched my example from several other implementations, and the reason behind `any` is exactly as you say. Maybe @jen20 can clarify how `Input`s get resolved and where we stand on using types everywhere?

praneetloke replied:
Yeah I wonder if I could do a union type here for my args like: `pulumi.Input<T> | string`, and then still be able to pass an `Input<T>` from the outside, but within my dynamic resource use them as regular strings?

mikhailshilkov replied:
So what we can do is to create two distinct types: an internal `Props` (name not final) which has no inputs in it and an exported `Args` which has the same shape but has all inputs. `Props` will be used in the dynamic provider implementation and `Args` is like normal args in all our components - to be passed into the constructor.

In order to avoid copy-pasting and typos, we can define a mapped type. I tried it on my static website example:

``` ts
// A mapped type to convert any interface into a type of the same shape but all types converted from T to Input<T>
type Inputs<T> = {
    [P in keyof T]: pulumi.Input<T[P]>;
}
  
interface StorageStaticWebsiteProps {
    accountName: string;
}

export type StorageStaticWebsiteArgs = Inputs<StorageStaticWebsiteProps>;
```

I was able to replace all `any` with `StorageStaticWebsiteProps`. Works like a charm.

@CyrusNajmabadi @lukehoban Don't we have a type like `Inputs` above in the standard Pulumi library? Do we what one?

praneetloke replied:
This is an elegant solution and one that I prefer, but would actually love it if it was part of the dynamic providers library itself. Aside from that, I do feel mapped types are sort of an advanced concept in TS for most beginners who are using TS, let alone Pulumi. That's just my opinion, though. I'd love to see what @lukehoban thinks about this.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+});
+
+// Create a Blob container in the storage account.
+const blobContainer = new azure.storage.Container("blob-container", {
```
Consider documenting (in README and/or inline comments) where the static website files must be put and which URL to use to access them through CDN.

praneetloke replied:
Will do.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+const location = "West US";
+
+/** The custom domain host of the CDN endpoint. */
+export let cdnCustomDomainResource: CDNCustomDomainResource | undefined;
```
What is the reason to export it?

praneetloke replied:
To show that the dynamic resource's outputs are exportable too.

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+
+import { CDNCustomDomainResource } from "./cdnCustomDomain";
+
+const location = "West US";
```
Consider using a configuration parameter for the location instead of a hard-coded value.

praneetloke replied:
I'll leave it as-is, but will add a comment that the user should consider using the Pulumi config system to externalize it.

mikhailshilkov replied:
from @ericrudder in another example:
>it woudl be great if we could "unhard code these" everywhere ... or at least not grow debt
> it's really a pulumi aniti-pattern, and we should avoid these ... we have a few others (like poor password usage) and they bite us when users copy/paste

mikhailshilkov commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,86 @@
+const cdnProfile = new azure.cdn.Profile("cdn-profile", {
+    location: location,
+    resourceGroupName: resourceGroup.name,
+    sku: "Standard_Akamai",
```
What are the implications of choosing Akamai instead of Microsoft?

praneetloke replied:
I'll add a link to https://docs.microsoft.com/en-us/azure/cdn/cdn-features.

mikhailshilkov commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,231 @@
+        inputs.endpointName,
+        inputs.customDomainName,
+        inputs.customDomainHostName);
+    console.log(`Custom domain provisioning state is ${ result.resourceState }.`);
```
Do we want to keep `console.log` in the examle?

praneetloke replied:
I'd like to since the custom domain and HTTPS provisioning can, in some cases, not be immediately active.

praneetloke commented:
> This seems to be related to #306, although with a different focus. Is that something we could/should combine?

I somehow missed this comment of yours and see that your PR is now merged. In any case, FWIW, I would have replied that I would not like to merge the two examples. I'd like to keep them as distinct ones so it doesn't look like one large example. Though, I do agree that the two go together very well.

mikhailshilkov commented:
I'm getting an error during `cdnCustomDomain` preview/creation:

```
Error: Error "SyntaxError: Unexpected token o in JSON at position 1" occurred while parsing the response body - [object Object].: Error: Error "SyntaxError: Unexpected token o in JSON at position 1" occurred while parsing the response body - [object Object].
        at new RestError (C:\work\pulumi\examples\azure-ts-dynamicresource\node_modules\@azure\ms-rest-js\lib\restError.ts:18:5)
        at errorHandler (C:\work\pulumi\examples\azure-ts-dynamicresource\node_modules\@azure\ms-rest-js\lib\policies\deserializationPolicy.ts:196:15)
        at process._tickCallback (internal/process/next_tick.js:68:7)

  pulumi-nodejs:dynamic:Resource (azure:cdn:Endpoint:CustomDomains:cdnCustomDomain):
    error: Error "SyntaxError: Unexpected token o in JSON at position 1" occurred while parsing the response body - [object Object].
```

Running from CLI, other resources are already created. Any ideas?

mikhailshilkov commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,244 @@
+  httpsEnabled: boolean;
+}
+
+export interface CustomDomainOutputs extends cdnManagement.CdnManagementModels.CustomDomainsCreateResponse {
```
Is this an internal structure to bring the state between multiple runs or will a user interact with it? If the former, why is it exported? If the latter, its shape seems to be different from what a typical component would expose as outputs? Is there guidance for dynamic providers in this regards?

praneetloke replied:
> Is this an internal structure to bring the state between multiple runs or will a user interact with it? If the former, why is it exported?

I exported it because I am under the assumption that the type of the outputs objects could be used by the caller to extract specific properties. But I am not sure if it is necessary, and if there is a more "accepted" way for dynamic providers to declare an output type or if one is even needed. At the very least I could un-export this type.

> Is there guidance for dynamic providers in this regards?

I am going to speak to someone here at Pulumi about dynamic provider outputs too and have an answer for this.

praneetloke commented:
> Running from CLI, other resources are already created. Any ideas?

Yes, see https://github.com/Azure/azure-sdk-for-js/issues/2842#issuecomment-498895303. MS just fixed an issue that crept into their SDK when they upgraded `axios` to `0.19.0` to patch a vulnerability.

mikhailshilkov commented on azure-ts-dynamicresource/package.json:
```diff
@@ -0,0 +1,13 @@
+    },
+    "dependencies": {
+        "@azure/arm-cdn": "^4.2.0",
+        "@azure/ms-rest-nodeauth": "^1.1.1",
```
Shall we set the versions to the ones that don't cause the error?

praneetloke replied:
Done.

CyrusNajmabadi commented on azure-ts-dynamicresource/README.md:
```diff
@@ -0,0 +1,85 @@
+            // to automatic renewal at no additional cost to you.
+            httpsEnabled: true,
+            resourceGroupName: resourceGroupName
+        }, { parent: cdnEndpoint });
```
nit: we would dedent this all twice.

CyrusNajmabadi replied:
not sure why this was resolved.

praneetloke replied:
Oh oops. Didn't mean to.

praneetloke replied:
It's updated now.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,244 @@
+import * as pulumi from "@pulumi/pulumi";
```
copyright?

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,244 @@
+import { ServiceClientCredentials } from "@azure/ms-rest-js";
+
+export interface CustomDomainOptions {
+  resourceGroupName: string;
```
nit: you have 2-space indent on.  But i think all our sdk/example code is 4-space.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,244 @@
+
+    return result;
+  }
+  async diff(id: string, previousOutput: CustomDomainOutputs, news: CustomDomainOptions): Promise<pulumi.dynamic.DiffResult> {
```
newline above this.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,244 @@
+      outs: currentOutputs
+    };
+  }
+}
```
note: i actually skipped most of this as i don't feel qualified to review a provider.  perhaps w can get @pgavlin to take a look.  It's something i want to ramp up on, so i'll read this and ask questions.  But i'd like someone more informed to take a deeper look.

CyrusNajmabadi commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,96 @@
+import * as pulumi from "@pulumi/pulumi";
```
copyright.

CyrusNajmabadi commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,96 @@
+    isHttpAllowed: false,
+    isCompressionEnabled: true,
+    originHostHeader: storageAccount.primaryBlobHost,
+    contentTypesToCompresses: [
```
aside: contentTypesToCompresses is a bizarro name.  "Compresses"??

praneetloke replied:
Yes. Unfortunately, that's how the terraform property is named. It is a typo in the provider itself.

CyrusNajmabadi commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,96 @@
+            // The certificate is fully managed by Azure from provisioning
+            // to automatic renewal at no additional cost to you.
+            httpsEnabled: true,
+        }, { parent: cdnEndpoint });
```
i'm a little confused why any of this would be needed.  can't you just write:

```ts
cdnCustomDomainResource = new ...(... {
    resourceGroupName: resourceGroup.name,
    // etc.?
}
```

I don't see why you need the .all or .apply.

praneetloke replied:
The reason is it isn't clear from the docs about the dynamic provider if one can pass `Input<T>` props to the dynamic resource and still get strong typing in the provider's functions for the inputs.

praneetloke commented:
I just spoke to @pgavlin about the dynamic provider and I am going to make a bunch of changes based on that, as well as update our dynamic providers to reflect the recommendations.

CyrusNajmabadi commented:
Great.  Feel free to ping again when ready!

praneetloke commented:
@mikhailshilkov @CyrusNajmabadi @pgavlin can you please take another look now? I have cleaned-up a lot of things. The main difference is, I am now accepting `Input<T>` for the appropriate properties. Declared a separate interface that is local to the dynamic provider itself so I still get strong typing when dealing with the inputs in the CRUD functions. This is what was recommended by Pat.

Also, as recommended, I have declared the outputs as `public readonly` members with an `Output<T>` type in the `dynamic.Resource` class itself.

I will make an update to our dynamic providers docs based on this. I have added comments for most things in this PR.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+                        property: this.name,
+                        reason: validationOutput.reason || "domain_validation_failed"
+                    }
+                ]
```
consider trimming this down to:

```ts
            result = {
                failures: [{
                    property: this.name,
                    reason: validationOutput.reason || "domain_validation_failed",
                }],
```

(also in a few other cases).

praneetloke replied:
Done (and in a few other places)

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+        } else {
+            result = {
+                inputs: news
+            };
```
single prop objects are commonly written as `result = { inputs: news };`

CyrusNajmabadi replied:
assigning to result just to return below seems unnecessary.  you can just `return { inputs: news };`

praneetloke replied:
Sure. I can change it.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+            deleteBeforeReplace: deleteBeforeReplace,
+            replaces: replaces,
+            changes: changes
+        };
```
can be simplified to just: `return { deleteBeforeReplace, replaces, changes };`

praneetloke replied:
Really? Those are keys in the return type. Could I still do what you are suggesting?

CyrusNajmabadi replied:
yup.  in js/ts  `{ a, b, c }` is shorthand for `{ a: a, b: b, c: c }`.

praneetloke replied:
Ohh...learned something new!

praneetloke replied:
However, I'll leave it as-is in this case.

mikhailshilkov replied:
Why? I agree it reads redundant.

praneetloke replied:
I'll leave it up to the reader to make any such changes since it is more of a code style optimization. I don't think the example needs to be as succinct as possible.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+                    inputs.profileName,
+                    inputs.endpointName,
+                    this.name,
+                    inputs.customDomainHostName);
```
heavy wrapping.  consider just:

```ts
const result = await cdnClient.customDomains.create(
    inputs.resourceGroupName,
    inputs.profileName,
    inputs.endpointName,
    this.name,
    inputs.customDomainHostName);
```

Or, if it's not too long:

```ts
const result = await cdnClient.customDomains.create(
    inputs.resourceGroupName, inputs.profileName, inputs.endpointName, this.name, inputs.customDomainHostName);
```

CyrusNajmabadi replied:
Also, in general, in pulumi resources and whatnot, we have 'name' come first.

praneetloke replied:
Sure I did this here and in a couple of other places.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+                    inputs.resourceGroupName,
+                    inputs.profileName,
+                    inputs.endpointName,
+                    this.name);
```
same comments for this.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+        };
+
+        return {
+            id: result.id!,
```
not clear why the `!` is necessary.

praneetloke replied:
It is an optional in the `result`. I get a lint error for not checking for null-ness otherwise. Actually, I have a question for @pgavlin here. Should I be setting the value of the `id` in the output returned by the `create()`?

lukehoban replied:
> Should I be setting the value of the id in the output returned by the create()?

Yes - you'll want to ensure you return an `id` that represents the resource in the backing cloud provider.  This may be necessary for example to `delete` the resource later - as you'll need to use this `id` to look it up.

It's slightly concerning that the `id` is an optional return value from the cloud provider!

praneetloke replied:
> It's slightly concerning that the id is an optional return value from the cloud provider!

Yes, agreed. Though, it seems that the custom domain resource does always get an ID. I believe this is an optional because the underlying CDN management client extends itself from another type `Resource`, which is where this `id` comes from and it would seem that it is optional for other resources. So I think we are safe in this instance.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+        if (!result.resourceState) {
+            return;
+        }
+        if (result.resourceState !== 'Deleting') {
```
can you have a blank line above this for consistency.  or remove hte blank line between the first two ifs.

praneetloke replied:
Done.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+                    newInputs.resourceGroupName,
+                    newInputs.profileName,
+                    newInputs.endpointName,
+                    this.name);
```
same comments about wrapping.

CyrusNajmabadi commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,276 @@
+                newInputs.resourceGroupName,
+                newInputs.profileName,
+                newInputs.endpointName,
+                this.name);
```
same comment.

CyrusNajmabadi commented on azure-ts-dynamicresource/index.ts:
```diff
@@ -0,0 +1,95 @@
+const location = "West US";
+
+/** The custom domain host of the CDN endpoint. */
+export let cdnCustomDomainResource: CDNCustomDomainResource | undefined;
```
why declare this up here.  you can just move down to the place you actually create it.

praneetloke replied:
Yep. Moved it.

mikhailshilkov commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,258 @@
+ * to each function in the implementation of a `pulumi.dynamic.ResourceProvider`.
+ * The property names in this must match `CustomDomainOptions`.
+ */
+interface DynamicProviderInputs {
```
Don't you like my suggested trick with a mapped type?

praneetloke replied:
Ah it got buried under my updates. I'll reply to that comment.

pgavlin commented on azure-ts-dynamicresource/README.md:
```diff
@@ -0,0 +1,85 @@
+
+## Prerequisites
+
+Prior to running this example, ensure that the custom domain, you will be adding to your CDN endpoint, has a CNAME record in your domain registrar's DNS management portal. The dynamic provider will validate that the mapping exists and will fail the deployment if the CDN endpoint is not reachable through your custom domain.
```
nit: extra commas

> Prior to running this example, ensure that the custom domain you will be adding to your CDN endpoint has a CNAME record in your domain registrar's DNS management portal

pgavlin commented on azure-ts-dynamicresource/README.md:
```diff
@@ -0,0 +1,85 @@
+
+## Known Issues
+
+If you get a 404 error when deleting a custom domain or a deserialization error `SyntaxError: Unexpected token o in JSON at position 1`, this was due to a bug in the Azure nodeJS SDK. See https://github.com/Azure/azure-sdk-for-js/issues/2842. Ensure you have the latest versions by running `npm list @azure/ms-rest-js @azure/ms-rest-azure-js`. The versions you should have are:
```
reword

> Bugs in older versions of the Azure NodeJS SDK can cause 404 errors or deserialization errors during some operations (see https://github.com/Azure/azure-sdk-for-js/issues/2842 for more details). You can ensure that you have the latest versions of the relevant SDK packages by running...

praneetloke replied:
Reworded it differently.

pgavlin commented on azure-ts-dynamicresource/README.md:
```diff
@@ -0,0 +1,85 @@
+- `@azure/ms-rest-js` - 1.8.10
+- `@azure/ms-rest-azure-js` - 1.3.7
+
+If you do not have the latest versions, then delete the `node_modules` folder, and the `package-lock.json`, or if you are using yarn, then `yarn-lock.json`, and then run `npm install` (or `yarn install`) again. Then verify that you have the correct versions again by running the `npm list ...` command.
```
> If you do not have the latest versions, delete your `node_modules` folder and `package-lock.json` (or `yarn.lock` if you are using yarn) and run `npm install` (or `yarn install`) again.

praneetloke replied:
Done.

pgavlin commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,258 @@
+    profileName: pulumi.Input<string>;
+    endpointName: pulumi.Input<string>;
+    customDomainHostName: pulumi.Input<string>;
+    httpsEnabled: boolean;
```
`pulumi.Input<boolean>`?

praneetloke replied:
I intentionally left it since its value is derived from a pulumi config property. Would you still recommend that I have it as an `Input<boolean>`? How about a union type `Input<boolean> | boolean`?

mikhailshilkov replied:
`Input` is itself a union: `Input<T> = T | Promise<T> | OutputInstance<T>` so `Input<boolean>` includes `boolean`.

praneetloke replied:
Ah yes I see it now. I've changed it.

pgavlin commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,258 @@
+        let deleteBeforeReplace = false;
+
+        if (previousOutput.customDomainHostName !== news.customDomainHostName || previousOutput.name !== this.name) {
+            console.warn("Changing the domain name properties will cause a downtime.")
```
`s/cause a downtime/cause downtime`

pgavlin commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,258 @@
+
+        // HTTPS can be enabled/disabled in-place.
+        if (previousOutput.httpsEnabled !== news.httpsEnabled) {
+            changes = true;
```
Changes really ought to be true in all cases that there are changes, even if those changes are already listed in `replaces`.

praneetloke replied:
Ah gotcha I'll change it. Reading the doc for [`diff`](https://pulumi.io/reference/programming-model/#diff-id-olds-news) again, I now see what it means, that it is regardless of the type of update; replace vs delete/create.

pgavlin commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,258 @@
+            throw new Error(`Provisioning state of the custom domain was expected to be 'Deleting', but was ${result.resourceState}.`);
+        }
+
+        console.log("The request to delete was successful. However, it can take a minute or two to fully complete deletion.");
```
FWIW, the console logging is pretty unusual for a provider. I think it makes more sense for these to be ephemeral log messages (like the k8s status messages), but I can't recall offhand how to do that.

praneetloke replied:
`pulumi.log.info()|warn()|...`? I see that those functions accept an ephemeral property. It'll be good to demonstrate that to the reader, that Pulumi has that capability. I'll switch these to that.

mikhailshilkov commented on azure-ts-dynamicresource/cdnCustomDomain.ts:
```diff
@@ -0,0 +1,258 @@
+            return { inputs: news };
+        }
+
+        const validationOutput = await cdnClient.endpoints.validateCustomDomain(
```
When I run preview on a clean stack, I get an error (probably, because the resource group doesn't exist yet):

```
    Error: Resource group '04da6b54-80e4-46f7-96ec-b56ff0331ba9' could not be found.: Error: Resource group '04da6b54-80e4-46f7-96ec-b56ff0331ba9' could not be found.
        at new RestError (C:\work\pulumi\examples\azure-ts-dynamicresource\node_modules\@azure\ms-rest-js\lib\restError.ts:18:5)
        at C:\work\pulumi\examples\azure-ts-dynamicresource\node_modules\@azure\ms-rest-js\lib\policies\deserializationPolicy.ts:117:27
        at process._tickCallback (internal/process/next_tick.js:68:7)
```

praneetloke replied:
Hmm..does it pass when you re-run it?

praneetloke replied:
Something is not quite right here. I am looking into this.

praneetloke replied:
I have updated the provider to address this. The cause of this error was because `check` seems to be called during a preview as well, when the resource group doesn't exist yet. That is unfortunate, because the call to the validation of the custom domain seemed appropriate to have in the `check` function. I have moved it to the `create` function now. The `check` function is now just validating that there is a value for the required inputs.

When I used this for my own project originally, it was an existing app to which I added the custom domain, so the resource group et al. existed already, which is why I didn't notice this.

praneetloke commented:
I am planning to merge tomorrow, if there are no more comments on the core dynamic provider itself.

pgavlin commented on azure-ts-dynamicresource/README.md:
```diff
@@ -0,0 +1,85 @@
+
+## Known Issues
+
+If you get a 404 error when deleting a custom domain or a deserialization error `SyntaxError: Unexpected token o in JSON at position 1`, these were due to bugs in the older versions of the Azure nodeJS SDK. See https://github.com/Azure/azure-sdk-for-js/issues/2842 for more info. Ensure that you have the latest versions by running `npm list @azure/ms-rest-js @azure/ms-rest-azure-js`. The versions  ...
```
nit: `NodeJS`
